src/nyx/gpu/image_moments.cu

```diff
@@ -7,56 +7,87 @@
 #include "../features/image_matrix.h"
 #include "gpu.h"
 
-static const int blockSize = 512;
+static const int NUM_BLOCKS = 16; // used by reducing kernels
+static const int blockSize = 512;   // used by non-reducing kernels
 
-size_t* devIndices = nullptr;
-StatsInt* devContourData = nullptr;
+namespace Nyxus
+{
+    // Image matrices (allocated and initialized with data each time a pending ROI batch is formed)
+    extern PixIntens* ImageMatrixBuffer;
+    PixIntens* devImageMatrixBuffer = nullptr;

```

Comment:
Is it possible to use std::vectors here instead of pointers?

Reply:
In general avoiding raw pointers is a good practice, but a pair of dual CPU- and GPU-side buffers like "ImageMatrixBuffer" and "devImageMatrixBuffer" is not a case where a raw pointer can be plainly represented by std::vector(s). The reasons are the following. Ignoring the fact that then there would exist a wasted allocation of initial small amount of memory at the vector construction in the global scope that we'd later reallocate to a desired memory size, yes, we can do so to assure deallocation of the host-side memory buffer "ImageMatrixBuffer" when the vector instance goes out of scope. But we cannot do a similar thing to ImageMatrixBuffer's device-side counterpart - "devImageMatrixBuffer" which needs to be allocated via CUDA API. Using std::vector's buffer for hosting only the host-side memory would break the host/device buffer symmetry. Another problem with the idea of implicit using std::vector's memory allocation mechanism in this context is inability to nicely lock the vector instance for accidental implicit resizing its internal buffer as a result of push_back() between its allocation and disposing, while a similar accidental resizing a raw pointer defined buffer is not possible.